nav2_behavior_tree/plugins/action/truncate_path_local_action.cpp

```diff
@@ -0,0 +1,178 @@
+#include "nav2_util/robot_utils.hpp"
+#include "nav_msgs/msg/path.hpp"
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_behavior_tree/plugins/action/truncate_path_local_action.hpp"
+
+namespace nav2_behavior_tree
+{
+
+TruncatePathLocal::TruncatePathLocal(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::ActionNodeBase(name, conf)
+{
+  tf_buffer_ =
+    config().blackboard->template get<std::shared_ptr<tf2_ros::Buffer>>(
+    "tf_buffer");
+}
+
+inline BT::NodeStatus TruncatePathLocal::tick()
+{
+  setStatus(BT::NodeStatus::RUNNING);
+
+  double distance_forward, distance_backward;
+  geometry_msgs::msg::PoseStamped pose;
+  double angular_distance_weight;
+  double max_robot_pose_search_dist;
+
+  getInput("distance_forward", distance_forward);
+  getInput("distance_backward", distance_backward);
+  getInput("angular_distance_weight", angular_distance_weight);
+  getInput("max_robot_pose_search_dist", max_robot_pose_search_dist);
+
+  bool path_pruning = std::isfinite(max_robot_pose_search_dist);
+  updatePath(path_pruning);
+
+  if (!getRobotPose(path_.header.frame_id, pose)) {
+    return BT::NodeStatus::FAILURE;
+  }
+
+  if (path_.poses.empty()) {
+    setOutput("output_path", path_);
+    return BT::NodeStatus::SUCCESS;
+  }
+
+  auto closest_pose_detection_end = path_.poses.end();
+  if (path_pruning) {
+    closest_pose_detection_end = nav2_util::geometry_utils::first_after_integrated_distance(
+      closest_pose_detection_begin_, path_.poses.end(), max_robot_pose_search_dist);
+  }
+
+  // find the closest pose on the path
+  auto current_pose = nav2_util::geometry_utils::min_by(
+    closest_pose_detection_begin_, closest_pose_detection_end,
+    [&pose, angular_distance_weight](const geometry_msgs::msg::PoseStamped & ps) {
+      return poseDistance(pose, ps, angular_distance_weight);
+    });
+
+  if (path_pruning) {
+    closest_pose_detection_begin_ = current_pose;
+  }
+
+  // expand forwards to extract desired length
+  auto forward_pose_it = nav2_util::geometry_utils::first_after_integrated_distance(
+    current_pose, path_.poses.end(), distance_forward);
+
+  // expand backwards to extract desired length
+  // Note: current_pose + 1 is used because reverse iterator points to a cell before it
+  auto backward_pose_it = nav2_util::geometry_utils::first_after_integrated_distance(
+    std::reverse_iterator(current_pose + 1), path_.poses.rend(), distance_backward);
+
+  nav_msgs::msg::Path output_path;
+  output_path.header = path_.header;
+  output_path.poses = std::vector<geometry_msgs::msg::PoseStamped>(
+    backward_pose_it.base(), forward_pose_it);
+  setOutput("output_path", output_path);
+
+  return BT::NodeStatus::SUCCESS;
+}
+
+inline void TruncatePathLocal::updatePath(bool reset_path_pruning_on_change)
+{
+  if (reset_path_pruning_on_change) {
+    nav_msgs::msg::Path new_path;
+    getInput("input_path", new_path);
+
+    // detect for changes
+    bool changed = true;
```

Comment:
I believe just `new_path != path_ -> changed = true` would suffice. I believe that would iterate through the internal structures.

Reply:
The entire block from L109-121 can just be 

`if (new_path != path_) {path_ = new_path; closest_pose_detection_begin_ = path_.poses.begin();} else {...}`

Reply:
I feel like I'm missing something subtle here

Reply:
Yes, it can be simplified. I just wanted to leave out comparisons of headers, thinking that there's no need to reset pruning if poses don't actually change. But maybe I just overthinked it...